modules/tiles/src/tileset/tileset-3d.js

```diff
@@ -219,35 +221,113 @@ export default class Tileset3D {
-    // TODO: update traverser to work with multiple viewports. Take viewport[0] for now
-    if (viewport instanceof Array) {
-      viewport = viewport[0];
+  /**
+   * Update visible tiles relying on a list of deck.gl viewports
+   * @param {WebMercatorViewport} viewports - list of deck.gl viewports
+   * @return {void}
+   */
+  update(viewports) {
+    if (this.travererseCounter > 0) {
+      return;
+    }
+    if (!(viewports instanceof Array)) {
+      viewports = [viewports];
     }
+
     this._cache.reset();
     this._frameNumber++;
-    this._frameState = getFrameState(viewport, this._frameNumber);
-    this._traverser.traverse(this.root, this._frameState, this.options);
+    this.traverseCounter = viewports.length;
+    // First loop to decrement traverseCounter
+    for (const viewport of viewports) {
+      const id = viewport.id;
+      this._initTraverser(id);
+    }
+
+    // Second loop to traverse
+    for (const viewport of viewports) {
+      const id = viewport.id;
+      if (!this.roots[id]) {
+        this.roots[id] = this._initializeTileHeaders(this.tileset, null, this.basePath);
+      }
+
+      if (!this._traversers[id]) {
+        continue; // eslint-disable-line no-continue
+      }
+      const traverser = this._traversers[id];
+      const frameState = getFrameState(viewport, this._frameNumber);
+      traverser.traverse(this.roots[id], frameState, this.options);
+    }
+  }
+
+  /**
+   * Create a traverser for a viewport if necessary
+   * @param {string} viewportId - id of a viewport
+   * @return {TilesetTraverser}
+   */
+  _initTraverser(viewportId) {
+    let traverserId = viewportId;
+    if (this.options.viewportTraversersMap) {
+      traverserId = this.options.viewportTraversersMap[viewportId];
+    }
+    if (traverserId !== viewportId) {
+      this.traverseCounter--;
+      return null;
+    }
+
+    if (!this._traversers[traverserId]) {
+      this._traversers[traverserId] = this._initializeTraverser();
+    }
+    return this._traversers[traverserId];
+  }
+
+  /**
+   * The callback to post-process tiles after traversal procedure
+   * @param {TilesetTraverser} traverser - the TilesetTraverser instance that just finished traversal
+   * @param {Object} frameState - frame state for tile culling
+   * @return {void}
+   */
+  _onTraversalEnd(traverser, frameState) {
+    const id = frameState.viewport.id;
+    if (!this.frameStateData[id]) {
+      this.frameStateData[id] = {selectedTiles: [], _requestedTiles: [], _emptyTiles: []};
+    }
+    const currentFrameStateData = this.frameStateData[id];
+    const selectedTiles = Object.values(traverser.selectedTiles);
+    currentFrameStateData.selectedTiles = selectedTiles;
+    currentFrameStateData._requestedTiles = Object.values(traverser.requestedTiles);
+    currentFrameStateData._emptyTiles = Object.values(traverser.emptyTiles);
+
+    this.traverseCounter--;
+    if (this.traverseCounter > 0) {
+      return;
+    }
+
+    this._updateTiles();
```

Comment:
Can we move the `_updateTiles()` call to the caller of this function?

If so, can we drop the `traverseCounter`?

Reply:
This method (` _onTraversalEnd`) is callback. The caller is an instance of another class.